vassal-app/src/main/java/VASSAL/build/module/index/VassalMapQuadTree.java

```diff
@@ -0,0 +1,153 @@
+package VASSAL.build.module.index;
+
+import VASSAL.build.module.Map;
+import VASSAL.build.module.map.boardPicker.Board;
+import VASSAL.counters.GamePiece;
+import VASSAL.tools.qtree.QNode;
+import VASSAL.tools.qtree.QPoint;
+import VASSAL.tools.qtree.QuadTree;
+
+import java.awt.Point;
+import java.awt.Rectangle;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A Quadtree index of pieces on the playable area of the Map.
+ * Each node of the Quadtree holds an Arraylist of the Game Pieces at that Map location
+ *
+ * Pieces that do not reside on the actual map (i.e. in the edge buffer, or off map) will not be
+ * indexed and NOT be found by a Ranged GKC
+ *
+ */
+public class VassalMapQuadTree extends QuadTree {
+
+  /** The actual playable bounds of the map **/
+  private final Rectangle bounds;
+
+  /** Cross-reference for each piece to it's current location in the Quadtree to allow for quick updates as pieces move
+   * This index records the currently recorded location of each piece, key is Piece ID.
+   * */
+  java.util.Map<String, Point> pieces = new HashMap<>();
+
+  /**
+   * Create a QuadTree to cover the entire map
+   * @param map
+   */
+  public VassalMapQuadTree(Map map) {
+
+    super();
+
+    int minX = 9999999;
+    int minY = 9999999;
+    int maxX = 0;
+    int maxY = 0;
+
+    // Find the Playable bounds of the entire Map
+    for (final Board b : map.getBoards()) {
+      final Rectangle bounds = b.bounds();
+      if (bounds.x < minX)
+        minX = bounds.x;
+      if (bounds.y < minY)
+        minY = bounds.y;
+      if (bounds.x + bounds.width > maxX)
+        maxX = bounds.x + bounds.width;
+      if (bounds.y + bounds.height > maxY)
+        maxY = bounds.y + bounds.height;
+    }
+
+    // Add 100 pixel border for good measure
```

Comment:
(In theory you could check for Decks & At-Start stacks outside the limit & adjust)

Reply:
The Quadtree implementation I am using requires a pre-determined bounds to work on, it can't be extended.  

It is always possible that a piece may be placed outside the board area at any time later and this will require a full rebuild of the quadtree.

Currently, I am not including pieces outside the playable board area in a ranged GKC lookup on the basis that it doesn't make any sense to find these pieces. 

However, on second thoughts, this a mistake as it will break thinks like a GKC on X, Y -1000, -1000. or SumLocation() on a piece outside the map area.

If a new piece is added to the index that is outside the current bounds, I will need to make a copy of the Qtree with bigger bounds before adding the piece.